pkg/storegateway/bucket_index_reader.go

```diff
@@ -628,43 +631,44 @@ func (r *bucketIndexReader) preloadSeries(ctx context.Context, ids []storage.Ser
 func (r *bucketIndexReader) loadSeries(ctx context.Context, ids []storage.SeriesRef, refetch bool, start, end uint64, loaded *bucketIndexLoadedSeries, stats *safeQueryStats) error {
 	begin := time.Now()
 
-	b, err := r.block.readIndexRange(ctx, int64(start), int64(end-start))
+	reader, err := r.block.indexRangeReader(ctx, int64(start), int64(end-start))
 	if err != nil {
 		return errors.Wrap(err, "read series range")
 	}
-
-	stats.update(func(stats *queryStats) {
-		stats.seriesFetchCount++
-		stats.seriesFetched += len(ids)
-		stats.seriesFetchDurationSum += time.Since(begin)
-		stats.seriesFetchedSizeSum += int(end - start)
-	})
+	defer runutil.CloseWithLogOnErr(r.block.logger, reader, "loadSeries close range reader")
+	byteReader := &uvarintSequenceReader{r: reader, offset: start}
 
 	for i, id := range ids {
-		c := b[uint64(id)-start:]
-
-		l, n := binary.Uvarint(c)
-		if n < 1 {
-			return errors.New("reading series length failed")
+		size, err := byteReader.Uvarint(uint64(id))
+		if err != nil {
+			return err
 		}
-		if len(c) < n+int(l) {
+		seriesBytes := loaded.bytesPool.Get(int(size))
+		n, err := io.ReadFull(byteReader, seriesBytes)
+		if n != int(size) || (err != nil && errors.Is(err, io.ErrUnexpectedEOF)) {
 			if i == 0 && refetch {
-				return errors.Errorf("invalid remaining size, even after refetch, remaining: %d, expected %d", len(c), n+int(l))
+				return errors.Errorf("invalid remaining size, even after refetch, read %d, expected %d", n, size)
 			}
 
 			// Inefficient, but should be rare.
 			r.block.metrics.seriesRefetches.Inc()
-			level.Warn(r.block.logger).Log("msg", "series size exceeded expected size; refetching", "id", id, "series length", n+int(l), "maxSeriesSize", maxSeriesSize)
-
+			level.Warn(r.block.logger).Log("msg", "series size exceeded expected size; refetching", "series_id", id, "series_length", size, "max_series_size", maxSeriesSize)
 			// Fetch plus to get the size of next one if exists.
-			return r.loadSeries(ctx, ids[i:], true, uint64(id), uint64(id)+uint64(n+int(l)+1), loaded, stats)
+			return r.loadSeries(ctx, ids[i:], true, uint64(id), uint64(id)+binary.MaxVarintLen64+size+1, loaded, stats)
```

Comment:
I'm puzzled by this logic, even the one before this PR. Let's say we have to refetch an ID in the middle of the ids slice. Here we're just calling `loadSeries()` on a bytes range which starts from id and ends at the end of the current id, but we'll have to read more after that. Tell me what I'm missing :)

Reply:
I thought the same. But it's actually not that bad. 

We can't refetch a series in the middle of the slice because the series are sorted and they all fall within the initial `[start, end)` range. We are guaranteed that the start of each series is in that range. If that's the case, then only the last series may be underfetched. This also works because series byte ranges cannot overlap.

Reply:
👍